### PR TITLE
[clang] Reject incomplete type arguments for __builtin_dump_struct

### LINKED ISSUE #63506: Clang frontend C++ crash when __builtin_dump_struct is used on pointer to undefined type
https://godbolt.org/z/fYs6ffrPT

```C++
#include <stdio.h>

struct Undefined;

struct Defined {
    Undefined* undef;
};

int main() {
    Undefined* value = nullptr;

    __builtin_dump_struct(value, printf);
    return 0;
}
```
Compiling this program crashes the clang frontend. Replacing the type of `value` with `Defined` emits correct code.

### PR DESCRIPTION
**Pull request by zyn0217**
We used to assume that the CXXRecordDecl passed to the 1st argument always had a definition. This is not true since a pointer to an incomplete type was not excluded.

Fixes https://github.com/llvm/llvm-project/issues/63506

### CODE REVIEW

File: clang/lib/Sema/SemaChecking.cpp [hunk 1]
```diff
@@ -713,6 +713,11 @@ static ExprResult SemaBuiltinDumpStruct(Sema &S, CallExpr *TheCall) {
     return ExprError();
   }
   const RecordDecl *RD = PtrArgType->getPointeeType()->getAsRecordDecl();
+  if (!RD->isCompleteDefinition()) {
```
Comment (zygoloid) [inline clang/lib/Sema/SemaChecking.cpp:716]: You need to use `RequireCompleteType` here to properly trigger instantiation of templated classes and handle definition visibility checking for modules.

Comment (zyn0217) [reply clang/lib/Sema/SemaChecking.cpp]: Thank you! Added the template case to the test.

**Review by shafik — APPROVED** [review]
LGTM you should put the bug report in the problem description so it automatically links the PR and closes it when you merge it.

**Comment by zyn0217** [comment]
Thanks! Added the link.

**Pull request merged**